### ISSUE TITLE
How to implement zeroconf in android app

### ISSUE THREAD

**Issue by powerbling** [librespot-org/librespot-java#535]
Hi, I cannot manage to understand how to implement zeroconf in an android app.
I want the app to be the receiver for spotify connect, in order to use old devices as spotify connect speakers.
I tried editing [https://github.com/devgianlu/librespot-android](url), copying the code from `librespot-player`'s `Main.java` and even if the server starts (according to the logs) the device is not discovered from spotify, both on mobile and desktop.

This is my code:
```java
        @Override
        public void run() {
            Log.i(TAG, "Running");
            ZeroconfServer server;
            if (LibrespotHolder.hasServer()) {
                server = LibrespotHolder.getServer();
                Log.i(TAG, "Using saved server");
                if (server == null) throw new IllegalStateException();
            } else {
                try {
                    Session.Configuration conf = new Session.Configuration.Builder()
                            .setStoreCredentials(false)
                            .setCacheEnabled(false)
                            .build();

                    ZeroconfServer.Builder builder = new ZeroconfServer.Builder(conf)
                            .setPreferredLocale(Locale.getDefault().getLanguage())
                            .setDeviceType(Connect.DeviceType.SPEAKER)
                            .setDeviceId(null).setDeviceName("librespot-android").setListenAll(true);


                    server = builder.create();
                    Log.i(TAG, "Started server");

                    LibrespotHolder.set(server);

                    server.addSessionListener(new ZeroconfServer.SessionListener() {

                        {
                            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                                if (LibrespotHolder.hasPlayer()) LibrespotHolder.getPlayer().close();
                            }));
                        }

                        @Override
                        public void sessionClosing(@NotNull Session session) {
                            if (LibrespotHolder.hasPlayer()) LibrespotHolder.getPlayer().close();

                        }

                        @Override
                        public void sessionChanged(@NotNull Session session) {
                            PlayerConfiguration playerConfiguration = new PlayerConfiguration.Builder()
                                    .setOutput(PlayerConfiguration.AudioOutput.CUSTOM)
                                    .setOutputClass(AndroidSinkOutput.class.getName())
                                    .build();

                            LibrespotHolder.set(new Player(playerConfiguration, session));

                        }
                    });

                    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                        try {
                            server.closeSession();
                            server.close();
                        } catch (IOException ignored) {
                        }
                    }));
                } catch (IOException ex) {
                    Log.e(TAG, "Session creation failed!", ex);
                    handler.post(() -> callback.failedGettingReady(ex));
                    return;
                }
            }

            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.M) {
                while (!LibrespotHolder.getPlayer().isReady()) {
                    try {
                        //noinspection BusyWait
                        Thread.sleep(100);
                    } catch (InterruptedException ex) {
                        return;
                    }
                }
            } else {
                try {
                    LibrespotHolder.getPlayer().waitReady();
                } catch (InterruptedException ex) {
                    LibrespotHolder.clear();
                    return;
                }
            }

            handler.post(() -> callback.playerReady("Prova"));
        }
```

This is the function I edited the most, other changes I made to the code are adding a server and removing the session from the `LibrespotHolder` class and removing `LoginActivity` and all code that calls it.

This a part of the output log and it seems the server starts (running on pixel 5 emu):
`I/x.g.z.Zeroconf: Announced Service{alias='librespot-android', service='spotify-connect', port=45345, text={VERSION=1.0, CPath=/, Stack=SP}, addresses=[/192.168.232.2, /192.168.200.2], domain='.local', protocol='tcp', host='pBW7kTiLABY=.local'}.`

I should mention that running `librespot-player` from windows the service is discoverable in spotify.

**Comment by devgianlu** [comment]
It seems like Android is now able to resolve `.local` domains on versions prior to Android 13. Source: https://issuetracker.google.com/issues/140786115

**Comment by powerbling** [comment]
Hi, after some experimenting I managed to make my device discoverable as a spotify connect receiver following the methods in this [notebook](https://gist.github.com/Fornoth/562a2db691051c381a8f).
I used [RxDNSSD](https://github.com/andriydruk/RxDNSSD) to transmit the mDNS broadcast and [AndroidAsync](https://github.com/koush/AndroidAsync#androidasync-also-lets-you-create-simple-http-servers) to serve the basic http server required for the info parsing by spotifys api.
I imagine that the `ZeroconfServer` class could be patched and create a constructor that doesn't initialize the custom `Zeroconf` implementation.
The system-specific broadcasting could be managed by another library (in this case RxDNSSD) and the `ZeroconfServer` could manage only the http server and other librespot logic.

Let me know @devgianlu if this could be done. I am available to make the changes myself.

**Comment by devgianlu** [comment]
Having a system like this should not be a problem, we are already doing it for the audio sink and stuff that on Android must be done differently.

**Comment by powerbling** [comment]
I implemented the changes in a private clone of the repo, I'm now tinkering with it to see if it can work. Should I open a pull request then? What's the process?

**Comment by powerbling** [comment]
I tried but it doesn't work. It seems like android doesn't like the server implementation.
I'll try to implement a custom ZeroconfServer for android. Should I update on the progress here?

**Comment by devgianlu** [comment]
> Should I update on the progress here?

Sure, just make a PR when you're done and I'll review it *eventually*.

**Comment by powerbling** [comment]
I ended up forking librespot-android because I felt like the scope of the original repo was too outside of what I came up with.
The original was meant to be a demo of the functionality while I built mine to be released with the specific intent of implementing spotify connect capability on android.
My modifications can still be merged with the original repo if need be.
My repo: https://github.com/powerbling/librespot-connect-android
